app/services/foreman/plugin/role_lock.rb

```diff
@@ -0,0 +1,65 @@
+module Foreman
+  class Plugin
+    class RoleLock
+      attr_reader :plugin_id
+
+      def initialize(plugin_id)
+        @plugin_id = plugin_id
+      end
+
+      def register_role(name, permissions, role_registry)
+        Role.transaction do
+          role = process_role name, permissions
+          role.ignore_locking do |r|
+            r.add_permissions!(permissions)
+          end
+          role_registry.role_ids << role.id
+        end
+      end
+
+      def process_role(name, permissions)
+        role = Role.find_by :name => name
+        if role && role.permission_diff(permissions).empty?
+          role.update_attribute :origin, @plugin_id if role.origin.empty?
+        elsif role
+          role = rename_and_create role, name, permissions
+        else
+          role = create_plugin_role name, permissions
+        end
+        role
+      end
+
+      def rename_and_create(role, original_name, permissions)
+        rename_existing role, original_name
+        create_plugin_role original_name, permissions
+      end
+
+      def create_plugin_role(name, permissions)
+        role = Role.create_locked! :name => name, :origin => @plugin_id
+        role.ignore_locking do |r|
+          r.add_permissions!(permissions)
+        end
+        role
+      end
+
+      def rename_existing(role, original_name)
+        prefix = "Customized"
+        role_name = generate_name prefix, original_name
+        if Role.find_by(:name => role_name)
```

Comment:
Because I know that I will find at most 1 - names are unique. But I do not know, how long I will have to iterate to find unique name, that is why there is a while loop - I try to find a role with newly generated name and if I get none, I can use it to rename.